offchain-modules/packages/scripts/src/utils/cardano_batch_test.ts

```diff
@@ -0,0 +1,125 @@
+import * as CardanoWasm from '@emurgo/cardano-serialization-lib-nodejs';
+import { asyncSleep } from '@force-bridge/x/dist/utils';
+import { logger } from '@force-bridge/x/dist/utils/logger';
+import CKB from '@nervosnetwork/ckb-sdk-core';
+import { WalletServer, Seed, AddressWallet, ShelleyWallet } from 'cardano-wallet-js';
+import { JSONRPCClient } from 'json-rpc-2.0';
+import fetch from 'node-fetch/index';
+import { prepareCkbPrivateKeys, prepareCkbAddresses, burn } from './eth_batch_test';
+
+export async function cardanoBatchTest(
+  ckbPrivateKey: string,
+  WALLET_SERVER_URL: string,
+  ckbNodeUrl: string,
+  ckbIndexerUrl: string,
+  forceBridgeUrl: string,
+  adaForceBridgeAddr: string,
+  adaWalletMnemonic: string,
+  batchNum = 100,
+  lockAmount = 2000000,
+  burnAmount = '1990000',
+): Promise<void> {
+  logger.info('adaBatchTest start!');
+  const ckb = new CKB(ckbNodeUrl);
+
+  const client = new JSONRPCClient((jsonRPCRequest) =>
+    fetch(forceBridgeUrl, {
+      method: 'POST',
+      headers: {
+        'content-type': 'application/json',
+      },
+      body: JSON.stringify(jsonRPCRequest),
+      // id: 1,
+    }).then((response) => {
+      if (response.status === 200) {
+        // Use client.receive when you received a JSON-RPC response.
+        return response.json().then((jsonRPCResponse) => client.receive(jsonRPCResponse));
+      } else if (jsonRPCRequest.id !== undefined) {
+        return Promise.reject(new Error(response.statusText));
+      }
+    }),
+  );
+
+  logger.info('adaBatchTest adaWallet created');
+
+  const ckbPrivs = await prepareCkbPrivateKeys(batchNum);
+  const ckbAddresses = await prepareCkbAddresses(ckb, ckbPrivs, ckbPrivateKey, ckbNodeUrl, ckbIndexerUrl);
+
+  const passphrase = 'user_wallet_passphrase';
+  const adaWallet = await getUserWallet(WALLET_SERVER_URL, adaWalletMnemonic, passphrase);
+
+  const lockTxs = await lock(client, adaWallet, passphrase, ckbAddresses, adaForceBridgeAddr, lockAmount, 10 * 1000);
+  // await check(client, lockTxs, ckbAddresses, batchNum, adaTokenAddress);
```

Comment:
The CI failure reason is `generateBridgeOutNervosTransaction error Error: sudt amount is not enough!`

It means you don't have enough sudt ( which is Ada mirror  token on Nervos here ) to transfer out from Nervos.

In the log I found the mint tx have already been successful. There might be some mismatch between the mint tx and burn tx, either the sudt type or the recipient address.

The `check` uses the `getBridgeTransactionSummaries` method in rpc server to check whether the lock transaction succeeds. You should implement the function in RPC  server to make the history part in UI work. And we can take advantage of this in this test to check whether our lock tx succeeds.

Reply:
I have added the 'check' API to debug the issue of burn-tx generation failure, and this is what I understood.
The following lock txs are created successfully.

```
lock txs [
  'd13cbc237ef82abe3393f4829dff53e9353e7f8b4ceb1dc84bcf7d8e71233eca',
  '18e74814f9fed4bc6044aa3dbf8235adcb226e57cedb3d3b1c0e3f36f9f988c3',
  'b93e79e102d7d6d5cefb6803a2c1c2d170dcf8bea30c714c14247ccc68ef7ac6'
]
```

These get handled by the collector, which creates the mint Tx for each of these.
For example:

```
^[[36m[2021-12-15T15:29:01.158 collector DEBUG handlers/ckb.ts:749]^[[39m CkbHandler waitUntilCommitted tx 0x86e770e47b99f0208cb53237df0034056feb09994afcc3a87e730f21a413fe5d status: committed, index: 33

```

After the successful commit of Tx on Ckb chain, it is detected by all the components. (I even checked through debug statements that the parseMintTx API succeeds)

```
[2021-12-15T15:29:16.685 verifier5 INFO handlers/ckb.ts:265]^[[39m CkbHandler onBlock handle logs from 38 to 39, mint txs: ["0x86e770e47b99f0208cb53237df0034056feb09994afcc3a87e730f21a413fe5d"], burn txs: []
[2021-12-15T15:29:16.692 collector INFO handlers/ckb.ts:265]^[[39m CkbHandler onBlock handle logs from 38 to 39, mint txs: ["0x86e770e47b99f0208cb53237df0034056feb09994afcc3a87e730f21a413fe5d"], burn txs: []
[2021-12-15T15:29:16.739 verifier3 INFO handlers/ckb.ts:265]^[[39m CkbHandler onBlock handle logs from 38 to 39, mint txs: ["0x86e770e47b99f0208cb53237df0034056feb09994afcc3a87e730f21a413fe5d"], burn txs: []
[2021-12-15T15:29:08.838 watcher INFO handlers/ckb.ts:265] CkbHandler onBlock handle logs from 37 to 38, mint txs: ["0x86e770e47b99f0208cb53237df0034056feb09994afcc3a87e730f21a413fe5d"], burn txs: []
[2021-12-15T15:29:16.882 watcher INFO handlers/ckb.ts:265] CkbHandler onBlock handle logs from 38 to 39, mint txs: ["0x86e770e47b99f0208cb53237df0034056feb09994afcc3a87e730f21a413fe5d"], burn txs: []
```

But the 'check' API sees the status as 'Pending' and therefore does not proceed further. This means that some component on the CKB side fails to register the mint tx.

```
tx {
  txSummary: {
    fromAsset: { network: 'Cardano', ident: 'ada', amount: '2000000' },
    toAsset: {
      network: 'Nervos',
      ident: '0xb786f9a911ae012c2e9f107d2961e75404ea71187de9e53238c0f435a8aaeb70',
      amount: null
    },
    sender: '438684494645d851b26b592cf8d389b064f707fed303f59582cf96d2a613fd9f#0',
    recipient: 'ckt1qyqvz0plmj60wwfdpykueer3kelclrp3ng4qc9qxla',
    fromTransaction: {
      txId: 'd13cbc237ef82abe3393f4829dff53e9353e7f8b4ceb1dc84bcf7d8e71233eca',
      timestamp: '2021-12-15T06:26:47.102Z',
      confirmStatus: 0
    }
  },
  status: 'Pending'
}
```